### PR TITLE
docs: add readme

### PR DESCRIPTION
**Pull request by atian25**
##### Checklist

- [ ] `npm test` passes
- [ ] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows commit guidelines

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -30,7 +30,137 @@ $ npm i egg-schedule --save
...
+};
+
+exports.task = function* (ctx) {
+  yield ctx.service.db.cleandb();
+};
+```
+
+## Overview
+
+`egg-schedule` is for time-based scheduling, not interval-based scheduling.
```
Comment (atian25) [inline README.md]: 我们算 time-based 还是 interval-based ? 有点乱乱搞不清.
- interval 和 cron 都是定时执行的, 不是相同间隔
- 要不要增加一个内建的支持, 等上一个执行完后才执行下一个?

**Comment by atian25** [comment]
@mattma,  help to review if you have time~ thanks ^_^

File: README.md [hunk 2]
```diff
@@ -30,7 +30,158 @@ $ npm i egg-schedule --save
...
+  };
+};
+
+// {app_root}/app/schedule/other.js
+exports.schedule = {
+  type: 'custom',
+};
+```
+
+## Dynamic schedule
```
Comment (atian25) [inline README.md]: 动态创建方式这里的设计没太理解场景  @dead-horse

File: README.md (hunk 1 above)
Comment (fengmk2) [inline README.md]: > 要不要增加一个内建的支持, 等上一个执行完后才执行下一个?

这个估计需要的，其实都不想并行执行的吧。 cc @dead-horse

File: README.md [hunk 3]
```diff
@@ -30,7 +30,158 @@ $ npm i egg-schedule --save
...
+*    *    *    *    *    *
+┬    ┬    ┬    ┬    ┬    ┬
+│    │    │    │    │    |
+│    │    │    │    │    └ day of week (0 - 7) (0 or 7 is Sun)
+│    │    │    │    └───── month (1 - 12)
+│    │    │    └────────── day of month (1 - 31)
+│    │    └─────────────── hour (0 - 23)
+│    └──────────────────── minute (0 - 59)
+└───────────────────────── second (0 - 59, optional)
+```
```
Comment (fengmk2) [inline README.md]: 好详细，点赞。

File: README.md (hunk 2 above)
Comment (fengmk2) [inline README.md]: 这个应该是给一个集群里面，只允许某台服务器运行

File: README.md [hunk 4]
```diff
@@ -30,7 +30,158 @@ $ npm i egg-schedule --save
...
+  - worker: will execute in one random worker when schedule executed.
+  - all: will execute in all workers when schedule executed.
+
+You can extend it by: 
+
+```javascript
+// {app_root}/agent.js
+const SCHEDULE_HANDLER = Symbol.for('egg#scheduleHandler');
+
+module.exports = function(agent) {
```
Comment (fengmk2) [inline README.md]: 使用 arrow function 吧

File: README.md [hunk 5]
```diff
@@ -30,7 +30,158 @@ $ npm i egg-schedule --save
...
+
+// {app_root}/app/schedule/other.js
+exports.schedule = {
+  type: 'custom',
+};
+```
+
+## Dynamic schedule
+
+```
```
Comment (fengmk2) [inline README.md]: js

File: README.md (hunk 2 above)
Comment (atian25) [inline README.md]: 动态创建我理解的场景是启动时读取数据库，然后创建。但这里不是 generator function

File: README.md [hunk 6]
```diff
@@ -30,7 +30,158 @@ $ npm i egg-schedule --save
...
+Just put your jobs at `{app_root}/app/schedule`, one file as one job.  
+
+Job file should export `schedule` and `task` properties.
+
+## Task
+Task should be a generator function, and accept ctx as params.
+
+You can find schedule log at log file, which act like worker request, and contain:
+- ctx.method: `SCHEDULE`
+- ctx.path: `/__schedules/${schedulePath}`
```
Comment (atian25) [inline README.md]: 这个地方有没有问题？我印象中死马改过

**Comment by popomore** [comment]
需要放 Egg 文档里？

**Comment by atian25** [comment]
egg 回头可以加个内置插件的介绍列表，然后链接到这边？

**Comment by mattma** [comment]
@atian25 I will do it tonight. U.S time. :)

File: README.md (hunk 1 above)
Comment (popomore) [inline README.md]: 单进程做调度很麻烦，而且不容易实现，感觉还是不要这么复杂的架构了。需要就依赖外部调度，扩展也强，也能集群化。

Comment (dead-horse) [inline README.md]: 之前也讨论过类似的问题，最终的结论是 egg-schedule 暂时不做这些保证啊，只做任务分发，不保证任务一定完成以及任务唯一执行。

其实真正看我们要用到定时任务的地方都会是一些相对较简单的功能，清缓存、定时刷新等功能，在没提供 schedule 的时候我们也不会去考虑这么多的吧

File: README.md (hunk 6 above)
Comment (dead-horse) [inline README.md]: 恩，改成了 `/__schedules?path=xxx` 吧应该，看看代码

File: README.md (hunk 2 above)
Comment (dead-horse) [inline README.md]: 要做到那么动态也是可以改的，但是现在我还没遇到，大多都是一些根据配置或者根据环境来决定是不是应该开启定时任务的情况

**Comment by mattma** [comment]
@atian25 I have gone over the entire readme and made lots of modifications. I just pushed a new commit instead of commenting on PR. I hope that you are OK with it. Let me know if I have any questions.

I do have a question. This documentation, [task section](https://github.com/eggjs/egg-schedule/blob/docs/README.md#task), need some love as well, but I do not understand the explanation. If look at the example, `ctx.service`, that did not even include in the list below. Can you explain what is your thoughts?

```
You can find schedule log at log file, which act like worker request, and contain:

ctx.method: SCHEDULE
ctx.path: /__schedules/${schedulePath}
ctx.query: scheule config(type=worker&cron=*%2F5%20*%20*%20*%20*%20*)
```

File: README.md [hunk 7]
```diff
@@ -61,36 +61,38 @@ exports.task = function* (ctx) {
...
-
-Job file should export `schedule` and `task` properties.
+To setup a schedule task, simply create a job file in `{app_root}/app/schedule`. A file contains one job and should export `schedule` and `task` properties. The rule of thumbs is one job per file.
 
 ## Task
-Task should be a generator function, and accept ctx as params.
+
+Task is a generator function, and accept one parameter, `ctx`. ex: `exports.task = function* (ctx) { ... };`
 
 You can find schedule log at log file, which act like worker request, and contain:
```
Comment (atian25) [inline README.md]: @mattma  
- egg will start one `master` process, cpu count `worker` process, one `agent` process
- `egg-schedule` will schedule at `agent` process,  but execute at `worker` process (one or all due to schedule type) 
- so when trigger, `agent` will send a message to `worker`, then the worker treat it just like a normal user http request. then this fake user request is use `SCHEDULE` as http method, `/__schedule?path=${s}&${qs.stringify(schedule.schedule)}` as url property.
- see https://github.com/eggjs/egg-schedule/blob/master/app.js#L29

File: README.md [hunk 8]
```diff
@@ -61,36 +61,38 @@ exports.task = function* (ctx) {
...
+
 - ctx.method: `SCHEDULE`
 - ctx.path: `/__schedules/${schedulePath}`
```
Comment (atian25) [inline README.md]: should change to `/__schedule`

File: README.md [hunk 9]
```diff
@@ -61,36 +61,38 @@ exports.task = function* (ctx) {
...
 - ctx.query: `scheule config(type=worker&cron=*%2F5%20*%20*%20*%20*%20*)`
```
Comment (atian25) [inline README.md]: `path=${s}&${qs.stringify(schedule.schedule)}`
should change to `path=<schedule file name>&<schedule config>` ?

File: README.md [hunk 10]
```diff
@@ -105,50 +107,53 @@ Use [cron-parser](https://github.com/harrisiirak/cron-parser).
...
+
 Build-in support is:
-  - worker: will execute in one random worker when schedule executed.
-  - all: will execute in all workers when schedule executed.
 
-You can extend it by: 
+  - **worker**: will be executed in one random worker when schedule run.
+  - **all**: will be executed in all workers when schedule run.
+
+**Custom schedule by extending the base schedule**
```
Comment (atian25) [inline README.md]: custom schedule is not extend `worker` or `all`,  just add a new one, not inheritance relationship

File: README.md [hunk 11]
```diff
@@ -105,50 +107,53 @@ Use [cron-parser](https://github.com/harrisiirak/cron-parser).
...
-example:
+Example:
+
 ```javascript
-// exce task every 3 hours
+// To execute task every 3 hours
 exports.schedule = {
   type: 'worker',
-  cron: '0 0 */3 * * *',
+  cron: '0 0 */3 * * *'
```
Comment (atian25) [inline README.md]: @mattma we use eslint `comma-dangle` rule,  so the comma is needed at code. 
anyway, docs is not required

**Comment by atian25** [comment]
thanks for let me know how poor is my english, hahaha

File: README.md (hunk 1 above)
Comment (atian25) [inline README.md]: 嗯, 等有需求再说.

话说, 目前有可能存在 schedule 挤压导致 worker 都被占满无法服务真实用户的情况

**Comment by atian25** [comment]
@dead-horse 这个没发送成功什么情况下会发生?  是测试提早 close 的原因? 这里是 INFO 的没问题吧?

> 2016-08-11 11:48:28,601 INFO 98460 [egg-schedule] message egg-schedule:/Users/tz/Workspaces/eggjs/egg-schedule/test/fixtures/worker/app/schedule/interval.js did not sent

File: README.md (hunk 11 above)
Comment (mattma) [inline README.md]: No wonder that most of source code have trailing comma. Good to know. I will keep it in mind.

**Comment by mattma** [comment]
> thanks for let me know how poor is my english, hahaha

我在美国生活，天天都讲，有这个语言环境。如果 @atian25  在这里生活一段时间，一定比我说的还好。 :) 

而且在阿里，你们每个人都有一个两个字的古代大侠的名字，真是太 cool 了。 我羡慕呀。 *_@

File: README.md (hunk 11 above)
Comment (atian25) [inline README.md]: using trailing comma, so:
- when add item to array,  diff will only one line
- and git blame of line will not got wrong

File: README.md (hunk 9 above)
Comment (mattma) [inline README.md]: @atian25 I have pushed a new commit for **tasks** section. feel free to check it out.

File: README.md (hunk 10 above)
Comment (mattma) [inline README.md]: fixed. @atian25 Ready to be reviewed.

**Comment by atian25** [comment]
+1

**Comment by atian25** [comment]
does these explain of schedule principle should include in docs?

let reader know that task is execute at worker

**Comment by mattma** [comment]
> does these explain of schedule principle should include in docs? let reader know that task is execute at worker

What do you mean exactly? For the `**Custom schedule**` section? Could you revise what I have written so that I can get a sense of what you are trying to share with our readers?

**Comment by atian25** [comment]
>  `egg-schedule` will schedule at `agent` process,  but execute at `worker` process (one or all due to schedule type) 
> so when trigger, `agent` will send a message to `worker`

**Comment by dead-horse** [comment]
@atian25 agent close 之后会触发的吧

**Comment by atian25** [comment]
@mattma I had update the docs: https://github.com/eggjs/egg-schedule/pull/5/commits/7d79d3e59dcf33e7b1e24e0652f719918b569615 , please help to review and feel free to update my grammar.

**Comment by mattma** [comment]
@atian25 working on a couple of PR on [cors](https://github.com/eggjs/egg-cors) and [security](https://github.com/eggjs/egg-security), will review yours before I go to bed tonight. :)

**Comment by mattma** [comment]
@atian25 Just went over all the comments, I think I have addressed most of docs related comments.

**Comment by atian25** [comment]
+1

**Comment by atian25** [comment]
landed 1.1.0

**Comment by popomore** [comment]
文档也要发版本？

**Comment by atian25** [comment]
上一个 feature 没发.

**Pull request merged**